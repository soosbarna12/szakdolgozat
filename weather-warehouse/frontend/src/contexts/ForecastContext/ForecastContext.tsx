import React, { createContext, useEffect, useState } from "react";
import { ForecastContextType, ForecastLocationData, ForecastProviderProps } from "./ForecastContext.type";

export const ForecastContext = createContext<ForecastContextType>({
  location: { name: "", lat: 0, lon: 0 },
  setLocation: () => {},
});

export const ForecastLocationProvider = ({ children }: ForecastProviderProps) => {
  const [location, setLocation] = useState<ForecastLocationData>(() => {
    const storedLocation = localStorage.getItem("forecastLocation");
    const defaultLocation = { name: "", lat: 0, lon: 0 };
    try {
      return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
    } catch (error) {
      return defaultLocation;
    }
  });

  useEffect(() => {
    // Persist location to localStorage whenever it changes
    localStorage.setItem("forecastLocation", JSON.stringify(location));
  }, [location]);

  return (
    <ForecastContext.Provider value={{ location, setLocation }}>
      {children}
    </ForecastContext.Provider>
  );
};