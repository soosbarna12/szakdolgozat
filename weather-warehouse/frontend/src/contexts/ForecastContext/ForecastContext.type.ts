import { Dispatch, ReactNode, SetStateAction } from "react";

export interface ForecastLocationData {
  name: string;
  country?: string;
  lat: number;
  lon: number;
}

export interface ForecastContextType {
  location: ForecastLocationData;
  setLocation: Dispatch<SetStateAction<ForecastLocationData>>;
}

export interface ForecastProviderProps {
  children: ReactNode;
}