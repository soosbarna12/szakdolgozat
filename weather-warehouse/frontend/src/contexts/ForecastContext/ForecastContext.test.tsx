import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import React, { useContext } from 'react';
import { ForecastContext, ForecastLocationProvider } from './ForecastContext';

describe('contexts/ForecastContext', () => {
  const mockChildren = <div data-testid="mockChildren">Mock Children</div>;

  beforeEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

  it('matches ForecastLocationProvider component snapshot', () => {
    const { container } = render(
      <ForecastLocationProvider>{mockChildren}</ForecastLocationProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('provides default values when localStorage is empty', () => {
    render(
      <ForecastLocationProvider>{mockChildren}</ForecastLocationProvider>
    );

    const defaultLocation = { name: '', lat: 0, lon: 0 };
    expect(localStorage.getItem('forecastLocation')).toEqual(
      JSON.stringify(defaultLocation)
    );

    const context = screen.getByTestId('mockChildren').closest('div');
    expect(context).toBeInTheDocument();
  });

  it('loads values from localStorage when available', () => {
    const mockLocation = { name: 'New York', lat: 40.7128, lon: -74.006 };
    localStorage.setItem('forecastLocation', JSON.stringify(mockLocation));

    render(
      <ForecastLocationProvider>{mockChildren}</ForecastLocationProvider>
    );

    expect(localStorage.getItem('forecastLocation')).toEqual(JSON.stringify(mockLocation));
  });

  it('updates localStorage when location changes', () => {
    const TestComponent = () => {
      const { location, setLocation } = useContext(ForecastContext);

      return (
        <div>
          <button
            data-testid="updateLocationButton"
            onClick={() => setLocation({ name: 'Los Angeles', lat: 34.0522, lon: -118.2437 })}
          >
            Update Location
          </button>
          <span data-testid="locationName">{location.name}</span>
        </div>
      );
    };

    render(
      <ForecastLocationProvider>
        <TestComponent />
      </ForecastLocationProvider>
    );

    const updateButton = screen.getByTestId('updateLocationButton');
    fireEvent.click(updateButton);

    expect(localStorage.getItem('forecastLocation')).toEqual(
      JSON.stringify({ name: 'Los Angeles', lat: 34.0522, lon: -118.2437 })
    );
    expect(screen.getByTestId('locationName').textContent).toBe('Los Angeles');
  });

  it('handles invalid JSON in localStorage gracefully', () => {
    localStorage.setItem('forecastLocation', 'invalid-json');

    render(
      <ForecastLocationProvider>{mockChildren}</ForecastLocationProvider>
    );

    const defaultLocation = { name: '', lat: 0, lon: 0 };
    expect(localStorage.getItem('forecastLocation')).toEqual(
      JSON.stringify(defaultLocation)
    );
  });
});