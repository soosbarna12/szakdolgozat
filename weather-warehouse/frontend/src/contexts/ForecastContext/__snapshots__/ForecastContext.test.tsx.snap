// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contexts/ForecastContext matches ForecastLocationProvider component snapshot 1`] = `
<div>
  <div
    data-testid="mockChildren"
  >
    Mock Children
  </div>
</div>
`;
